src/ucp/core/ucp_ep.c

```diff
@@ -3339,6 +3319,139 @@ static ucs_status_t ucp_ep_query_sockaddr(ucp_ep_h ep, ucp_ep_attr_t *attr)
+                                             ucs_string_buffer_t *strb,
+                                             void *arg_ptr, uint64_t arg_u64)
+{
+    struct sockaddr sockaddr;
+
+    if (ucp_ep_vfs_read_remote_addr((ucp_ep_h)obj, strb, &sockaddr) == UCS_OK) {
+        ucp_vfs_read_port(&sockaddr, strb);
+    }
+}
+
+static void ucp_ep_vfs_read_local_addr_ip(void *obj, ucs_string_buffer_t *strb,
+                                          void *arg_ptr, uint64_t arg_u64)
+{
+    struct sockaddr sockaddr;
+
+    if (ucp_ep_vfs_read_local_addr((ucp_ep_h)obj, strb, &sockaddr) == UCS_OK) {
+        ucp_vfs_read_ip(&sockaddr, strb);
+    }
+}
+
+static void ucp_ep_vfs_read_local_addr_port(void *obj,
+                                            ucs_string_buffer_t *strb,
+                                            void *arg_ptr, uint64_t arg_u64)
+{
+    struct sockaddr sockaddr;
+
+    if (ucp_ep_vfs_read_local_addr((ucp_ep_h)obj, strb, &sockaddr) == UCS_OK) {
+        ucp_vfs_read_port(&sockaddr, strb);
+    }
+}
+
+void ucp_ep_vfs_init(ucp_ep_h ep)
+{
+    ucp_err_handling_mode_t err_mode;
+    ucp_ep_attr_t attr;
+    struct sockaddr *sockaddr;
+
+#if ENABLE_DEBUG_DATA
+    ucs_vfs_obj_add_dir(ep->worker, ep, "ep/%s", ep->name);
+    ucs_vfs_obj_add_ro_file(ep, ucs_vfs_show_memory_address, NULL, 0,
+                            "memory_address");
+#else
+    ucs_vfs_obj_add_dir(ep->worker, ep, "ep/%p", ep);
+#endif
+
+    ucs_vfs_obj_add_ro_file(ep, ucp_ep_vfs_show_peer_name, NULL, 0,
+                            "peer_name");
+
+    err_mode = ucp_ep_config(ep)->key.err_mode;
+    ucs_vfs_obj_add_ro_file(ep, ucs_vfs_show_primitive,
+                            (void*)ucp_err_handling_mode_names[err_mode],
+                            UCS_VFS_TYPE_STRING, "error_mode");
+
+    attr.field_mask = UCP_EP_ATTR_FIELD_REMOTE_SOCKADDR;
+    if (ucp_ep_query_sockaddr(ep, &attr) == UCS_OK) {
+        sockaddr = (struct sockaddr*)&attr.remote_sockaddr;
+        ucs_vfs_obj_add_ro_file(ep, ucp_ep_vfs_read_remote_addr_ip, NULL, 0,
+                                "remote_address/%s",
+                                ucs_sockaddr_address_family_str(
+                                        sockaddr->sa_family));
+        ucs_vfs_obj_add_ro_file(ep, ucp_ep_vfs_read_remote_addr_port, NULL, 0,
+                                "remote_address/port");
+    }
+    attr.field_mask = UCP_EP_ATTR_FIELD_LOCAL_SOCKADDR;
+    if (ucp_ep_query_sockaddr(ep, &attr) == UCS_OK) {
+        sockaddr = (struct sockaddr*)&attr.local_sockaddr;
+        ucs_vfs_obj_add_ro_file(ep, ucp_ep_vfs_read_local_addr_ip, NULL, 0,
+                                "local_address/%s",
+                                ucs_sockaddr_address_family_str(
+                                        sockaddr->sa_family));
+        ucs_vfs_obj_add_ro_file(ep, ucp_ep_vfs_read_local_addr_port, NULL, 0,
+                                "local_address/port");
+    }
```

Comment:
seems there is a code duplication between local and remote address
can we narrow down the difference to 
{ "local", offsetof(attr, local_sockaddr), UCP_EP_ATTR_FIELD_LOCAL_SOCKADDR }
vs.
{ "remote", offsetof(attr, remote_sockaddr), UCP_EP_ATTR_FIELD_REMOTE_SOCKADDR }